tests/test_code_references.py

```diff
@@ -0,0 +1,203 @@
+
+
+def test_function_references():
+    def x():
+        return "42"
+
+    def y():
+        return x()
+
+    assert get_references(y) == [x]
+
+    def y():
+        return oneline("Use a function in another module")
+
+    assert get_references(y) == [oneline]
+
+    def y():
+        return func_does_not_exist()  # noqa: F821
+
+    assert get_references(y) == ["func_does_not_exist"]
+
+
+def test_class_references():
+    class MyClass:
+        def __init__(self):
+            self.value = "42"
+
+        def log_val(self):
+            import logging
+
+            logging.log(self.value)
+
+    def x():
+        my_class = MyClass()
+        return my_class
+
+    assert get_references(x) == [MyClass]
+
+    def x():
+        builder = FlowBuilder()
+        builder.assign("cls", MyClass)
+        return builder
+
+    assert get_references(x) == [FlowBuilder, "assign", MyClass]
+
+
+def test_method_references():
+    class MyClass:
+        def __init__(self):
+            self.value = "42"
+
+        def log_val(self):
+            import logging
+
+            logging.log(self.value)
+
+    def x():
+        my_class = MyClass()
+        my_class.log_val()
+
+    # We don't get the method as a reference because class initialization
+    # is a function call and that incorrectly sets my_class as None.
+    assert get_references(x) == [MyClass, "log_val"]
+
+    def y(my_class):
+        my_class.log_val()
+
+    assert get_references(y) == ["log_val"]
+
+
+def test_references_with_qualified_names():
+    import multiprocessing
+
+    def x():
+        """This function tests IMPORT_FROM opcode"""
+        from multiprocessing.managers import SyncManager
+
+        return SyncManager()
+
+    assert get_references(x) == [multiprocessing.managers.SyncManager]
+
+    def x():
+        """This function tests LOAD_ATTR opcode"""
+        p = multiprocessing.managers.public_methods
+        return p(multiprocessing.managers.SyncManager)
+
+    assert get_references(x) == [
+        multiprocessing.managers.public_methods,
+        multiprocessing.managers.SyncManager,
+    ]
+
+    def x():
+        """This function tests LOAD_METHOD opcode"""
+        m = multiprocessing.managers.SyncManager
+        return m.start()
+
+    assert get_references(x) == [
+        multiprocessing.managers.SyncManager.start,
+    ]
+
+
+def test_multiple_references():
+    def get_val(dict, key):
+        return dict[key]
+
+    def enclosing_x():
+        free_var = "free_val"
+
+        def x(func_def_var="func_def_val"):
+            local_var = {"local_key": "local_val"}
+            local_val = get_val(local_var, "local_key")
+            cell_var = "cell_val"
+            import logging
+
+            logging.log(global_val, free_var, local_val, cell_var, func_def_var)
+            builder = FlowBuilder("x")
+            # more business logic
+            builder.build()
+
+            def inner_x():
+                import logging
```

Comment:
Yup, that doesn't matter. The purpose of the inner function was to test for cell variable. The `inner_x` function is also a constant for `enclosing_x` function which will be hashed separately, so any references inside `inner_x` will be taken care of then.

I think this would become apparent when we integrate code referencing logic with our hashing logic and tests for them.